### PR TITLE
Adding the sorting in the projects and volunteers page

### PR DESCRIPTION
**Pull request by bansalnagesh**
Adding the sorting in Projects and Volunteers Page

Needs UX fixes

Closes Issue #56

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -7,11 +7,16 @@ class ProjectsController < ApplicationController
...
+  helper_method :sort_direction
+
   def index
     params[:page] ||= 1
     @show_filter = true
     @show_search_bar = true
 
+    sort = params[:sort] ? "#{params[:sort]} #{sort_direction}" : "highlight DESC, COUNT(volunteers.id) DESC, created_at DESC"
+
+    print "Here I am #{sort}"
```
Comment (wilsonw926) [inline app/controllers/projects_controller.rb]: remove the print statement here

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -7,11 +7,16 @@ class ProjectsController < ApplicationController
     only: [:show],
     if: Proc.new { |c| c.request.format.csv? }
 
+  helper_method :sort_direction
```
Comment (sradu) [inline app/controllers/projects_controller.rb]: Please don't use helper methods in controllers.

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -7,11 +7,16 @@ class ProjectsController < ApplicationController
...
+
   def index
     params[:page] ||= 1
     @show_filter = true
     @show_search_bar = true
 
+    sort = params[:sort] ? "#{params[:sort]} #{sort_direction}" : "highlight DESC, COUNT(volunteers.id) DESC, created_at DESC"
```
Comment (sradu) [inline app/controllers/projects_controller.rb]: This seems like an SQL injection issue.

File: app/helpers/registrations_helper.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module RegistrationsHelper
+  def sortable(title, column, direction)
```
Comment (sradu) [inline app/helpers/registrations_helper.rb]: It's a global namescape, perhaps volunteers_sortable? Also, perhaps volunteers_sortable_select or whatever HTML element it is.

File: app/helpers/projects_helper.rb [hunk 5]
```diff
@@ -4,4 +4,14 @@ def project_row_class(project)
 
     project_class = 'border-2 border-orange-300 bg-orange-100' if project.highlight
   end
+
+  def projects_sortable(title, column, direction)
+    direction = sort_direction == "asc" ? "desc" : "asc"
```
Comment (sradu) [inline app/helpers/projects_helper.rb]: Don't use both " and ' in the same code. We use " for html, and ' inside ruby.

File: app/controllers/projects_controller.rb [hunk 6]
```diff
@@ -150,4 +155,8 @@ def ensure_owner_or_admin
         redirect_to projects_path
       end
     end
+
+    def sort_direction
+      %w[asc desc].include?(params[:direction]) ? params[:direction] : 'asc'
```
Comment (sradu) [inline app/controllers/projects_controller.rb]: for ease of reading uppercase ASC, DESC etc.

**Comment by sradu** [comment]
If @bansalnagesh fixes this, @adrianthedev could you please add the UI?

**Comment by bansalnagesh** [comment]
image

File: app/controllers/projects_controller.rb [hunk 7]
```diff
@@ -9,6 +9,8 @@ def index
     @show_filters = true
     @show_search_bar = true
 
+    sort = params[:sort] ? "#{sanitize_sort_column(params[:sort])} #{sanitize_sort_direction(params[:direction])}" : 'highlight DESC, COUNT(volunteers.id) DESC, created_at DESC'
```
Comment (sradu) [inline app/controllers/projects_controller.rb]: Instead of doing this can you have params[:sort] point to a key of predefined sort values?

Comment (sradu) [reply app/controllers/projects_controller.rb]: eg

```
sort_options = { name_up: ‘name ASC, created_at DESC’, name_down: ‘name DESC, created_at DESC’ }
```

File: app/views/projects/index.html.erb [hunk 8]
```diff
@@ -55,6 +55,18 @@
   </div>
 <% end %>
 
+<br/>
```
Comment (sradu) [inline app/views/projects/index.html.erb]: Can you post a snapshot of this looks like? It should on the same line next to search.

Comment (bansalnagesh) [reply app/views/projects/index.html.erb]: image

image

**Comment by adrianthedev** [comment]
Will take a look and add the UI in a few

**Comment by adrianthedev** [comment]
I pushed some changes to simplify things a bit. If we need more filters please let me know.

**Comment by sradu** [comment]
I feel off not knowing what it's sorted by. That should be visible by default. Why not a select?

File: app/helpers/application_helper.rb [hunk 9]
```diff
@@ -96,4 +96,38 @@ def skill_badges(items, limit: nil, color: 'indigo', title: title)
...
+      if params[:sort_by] == sort_by
+        new_params.delete :sort_by
+        # active state
+        classes += ' bg-gray-100 text-gray-900'
+      else
+        new_params[:sort_by] = sort_by
+      end
+    else
+      new_params[:sort_by] = sort_by
+    end
```
Comment (tiegz) [inline app/helpers/application_helper.rb]:
```suggestion
    case params[:sort_by]
    when sort_by
      new_params.delete :sort_by
      # active state
      classes += ' bg-gray-100 text-gray-900'
    else
      new_params[:sort_by] = sort_by
    end
```

File: app/controllers/projects_controller.rb [hunk 10]
```diff
@@ -152,4 +156,9 @@ def ensure_owner_or_admin
         redirect_to projects_path
       end
     end
+
+    def get_order_param
+      return 'created_at desc' if params[:sort_by] == 'latest'
+      return 'volunteers.count asc' if params[:sort_by] == 'volunteers_needed'
```
Comment (tiegz) [inline app/controllers/projects_controller.rb]: I would assume "volunteers_needed" would be `desc` so you could see the projects with the most demand. WDYT?

File: app/helpers/application_helper.rb (hunk 9 above)
Comment (adrianthedev) [reply app/helpers/application_helper.rb]: much cleaner. thank you!

File: app/controllers/projects_controller.rb (hunk 10 above)
Comment (adrianthedev) [reply app/controllers/projects_controller.rb]: I was thinking that "volunteers needed" == "project doesn't have enough volunteers" so, show the ones with the least number of volunteers. Do you think that's misleading somehow?

Comment (tiegz) [reply app/controllers/projects_controller.rb]: Ahh, nevermind, I'm wrong! I was thinking of `number_of_volunteers` and missed that this was `volunteers.count`

**Pull request merged**